services/mailer/mailer.go

```diff
@@ -147,42 +147,54 @@ type smtpSender struct{}
 func (s *smtpSender) Send(from string, to []string, msg io.WriterTo) error {
 	opts := setting.MailService
 
-	host, port, err := net.SplitHostPort(opts.Host)
-	if err != nil {
-		return err
+	var network string
+	var address string
+	if opts.Protocol == "smtp+unix" {
+		network = "unix"
+		address = opts.SmtpAddr
+	} else {
+		network = "tcp"
+		address = net.JoinHostPort(opts.SmtpAddr, opts.SmtpPort)
 	}
 
-	tlsconfig := &tls.Config{
-		InsecureSkipVerify: opts.SkipVerify,
-		ServerName:         host,
+	conn, err := net.Dial(network, address)
+	if err != nil {
+		return err
 	}
+	defer conn.Close()
 
-	if opts.UseCertificate {
-		cert, err := tls.LoadX509KeyPair(opts.CertFile, opts.KeyFile)
-		if err != nil {
-			return err
+	var tlsconfig *tls.Config
+	if opts.Protocol == "smtps" || opts.Protocol == "smtp+startls" {
+		tlsconfig = &tls.Config{
+			InsecureSkipVerify: opts.ForceTrustServerCert,
+			ServerName:         opts.SmtpAddr,
 		}
-		tlsconfig.Certificates = []tls.Certificate{cert}
-	}
 
-	conn, err := net.Dial("tcp", net.JoinHostPort(host, port))
-	if err != nil {
-		return err
+		if opts.UseClientCert {
+			cert, err := tls.LoadX509KeyPair(opts.ClientCertFile, opts.ClientKeyFile)
+			if err != nil {
+				return err
+			}
+			tlsconfig.Certificates = []tls.Certificate{cert}
```

Comment:
A NPE can happen here if a user does have client cert specified but not the correct protocol(should probably be warned about in setting/mailer.go)

Reply:
I'm confused where the NPE occurs; shouldn't `cert` be non-nil if `err != nil`?

Reply:
`tlsconfig` can be nil when the user specifies a incorrect protocol:

Ex:
https://go.dev/play/p/QxyGjgFlBFM

Reply:
You might be misreading the diff; this is inside the block where `tlsConfig` is defined unconditionally.

Reply:
Dammit, yup.